### PR TITLE
Implement state transactions

### PR DESCRIPTION
**Pull request by ultratwo**
This pull request implements transactions for the State, this should enable reverts to be implemented. I have done some rudimentary testing.

Transactions are not currently supported by the optimized state.

> Beware of bugs in the above code; I have only proved it correct, not tried it. -- Donald Knuth

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: src/ethereum/frontier/state.py [hunk 1]
```diff
@@ -35,6 +35,53 @@ class State:
...
+
+    Parameters
+    ----------
+    state : State
+        The state.
+    """
+    state._snapshots.pop()
+
+
+def abort_transaction(state: State) -> None:
```
Comment (voith) [inline src/ethereum/frontier/state.py:74]: I prefer the name `rollback_transaction`. Thoughts?

File: src/ethereum/frontier/trie.py [hunk 2]
```diff
@@ -168,6 +169,24 @@ class Trie(Generic[T]):
...
+    ----------
+    trie: `Trie`
+        Trie to copy.
+
+    Returns
+    -------
+    new_trie : `T`
+        A copy of the trie.
+    """
+    return Trie(trie.secured, trie.default, copy.copy(trie._data))
```
Comment (voith) [inline src/ethereum/frontier/trie.py:187]: Sweet!

File: src/ethereum/frontier/state.py (hunk 1 above)
Comment (ultratwo) [reply src/ethereum/frontier/state.py]: Agreed.

File: src/ethereum/frontier/state.py [hunk 3]
```diff
@@ -35,6 +35,53 @@ class State:
         default_factory=lambda: Trie(secured=True, default=None)
     )
     _storage_tries: Dict[Address, Trie[U256]] = field(default_factory=dict)
+    _snapshots: List[
+        Tuple[Trie[Optional[Account]], Dict[Bytes, Trie[U256]]]
+    ] = field(default_factory=list)
+
+
+def start_transaction(state: State) -> None:
```
Comment (SamWilsn) [inline src/ethereum/frontier/state.py:43]:
```suggestion
def begin_transaction(state: State) -> None:
```

the other functions match SQL, so why not this one!

**Review by SamWilsn — APPROVED** [review]

**Comment by voith** [comment]
I will get rid of the deepcopy method that I added after this PR is merged.

**Comment by SamWilsn** [comment]
If you wouldn't mind looking over this quickly @voith and @ultratwo, I rebased and removed the deepcopy implementation.

**Review by voith — APPROVED** [review]
LGTM!

**Comment by ultratwo** [comment]
Looks good!

**Pull request merged**